bl32: add secure interrupt handling in AArch32 sp_min

etienne-lms opened the pull request:
Propose a minimal secure interrupt service in sp_min for the AArch32
implementation.

Change-Id: I1f7d1dc7b21f6f90011b7f3fcd921e455592f5e7
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
my apologies, this change is buggy, I did not rebase it.
fix is coming...

dp-arm commented on bl32/sp_min/aarch32/entrypoint.S:
```diff
@@ -176,24 +176,28 @@ func handle_fiq
 	isb
 
-	and 	r0, r4, #SECURE
+	push	{r2}
```
Don't we need 8 byte alignment here? (push some other dummy reg as well)

etienne-lms replied:
true, i'll update. thanks.

etienne-lms commented:
Squashed + rebased + fixed from review comment (alignment issue on stack usage).

soby-mathew commented:
@etienne-lms , we would like to understand more on the use case for FIQ handler in SP_MIN. The original intention of SP_MIN was to serve as a minimal image for just PSCI. Adding the FIQ handler now increases the scope and we want to understand the direction we are heading with this.

Also, the function `sp_min_fiq` takes `security_state` as an argument. Is that added to follow exisiting function prototypes or does it mean you need to allow interrupts to be triggered when executing in Secure mode ?

Also we need to understand where the responsibility of programming the SCR.FIQ bit lies (platform layer or generic layer).

etienne-lms commented:
@soby-mathew, the idea is to allow one to use the sp_min as a very basic runtime secure service stack, but not only the PSCI stack. Such service would indeed be quite simple since operating mode of the sp_min. Among such basic services could be handling some specific event we don't want to rely on "less safe" nonsecure world: forcing a platform reset or wiping some data from in secure ram on specific HW event detection (i.e some hw tampering).
Maybe such would require a specific configuration directive ? (`SP_MIN_WITH_SECURE_ISR=y`)

`security_state` handling in proposed `sp_min_fiq()` only aims at complying with generic prototype of `plat_sp_min_interrupt_handler()`. It is not expected to be meaningful and could be arbitrary force when calling `plat_sp_min_interrupt_handler()`, instead of being relayed from the FIQ root handler.

Regarding the setting of SCR.FIQ, we planned to configure it from the platform, but it could be set from the sp_min boot entrypoints.

soby-mathew commented:
Thanks @etienne-lms. I think we can clarify the following regarding the interrupt handler :
* SP_MIN can only handle FIQs. 
* It cannot pre-empt the secure world and the FIQ will only ever interrupt when executing in Normal world. So the FIQ handler need not save the secure register context.
* The build flag (SP_MIN_WITH_SECURE_ISR) will wrap all the code related to this.

The userguide would need updation with details of the build flag.

> security_state handling in proposed sp_min_fiq() only aims at complying with generic prototype of plat_sp_min_interrupt_handler(). It is not expected to be meaningful and could be arbitrary force when calling plat_sp_min_interrupt_handler(), instead of being relayed from the FIQ root handler.

OK, we can force this flag to be `non-secure` from generic code.

> Regarding  the setting of SCR.FIQ, we planned to configure it from the platform, but it could be set from the sp_min boot entrypoints

Yes, preferably the SCR.FIQ bit is set at both cold boot and warm boot entrypoints depending on the build flag.

etienne-lms commented:
Ok @soby-mathew. Thanks for the feedback. I will update the P-R accordingly.

Since this is restricted to FIQ, I propose to rename the build directive to `SP_MIN_WITH_SECURE_FIQ`.

By the way, do you prefer `SP_MIN_WITH_SECURE_ISR=y|n` or `SP_MIN_WITH_SECURE_ISR=0|1` ?

dp-arm commented:
> By the way, do you prefer SP_MIN_WITH_SECURE_ISR=y|n or SP_MIN_WITH_SECURE_ISR=0|1
I would go with the latter form as it is more consistent with what we use currently for other options.

etienne-lms commented:
P-R updates squashed and rebased.

danh-arm commented on bl32/sp_min/sp_min_main.c:
```diff
@@ -207,3 +207,25 @@ void sp_min_warm_boot(void)
+void plat_sp_min_interrupt_handler(uint32_t id)
+{
+	VERBOSE("[sp_min] secure interrupt %d\n", (int)id);
+}
```
I don't see the value in this weak definition. The feature has no value unless the platform provides a useful implementation. I suggest removing to force the platform to provide an implementation.

etienne-lms replied:
Ok

danh-arm commented on include/bl32/sp_min/platform_sp_min.h:
```diff
@@ -17,4 +17,6 @@ void sp_min_plat_runtime_setup(void);
 void sp_min_plat_arch_setup(void);
 entry_point_info_t *sp_min_plat_get_bl33_ep_info(void);
 
+void plat_sp_min_interrupt_handler(uint32_t id);
```
The other functions in this file are called `sp_min_plat_xxx()`. You should align with this. Also wouldn't `sp_min_plat_fiq_handler()` be more accurate?

etienne-lms replied:
right and true. i will update. thanks.

danh-arm commented on include/common/aarch32/el3_common_macros.S:
```diff
@@ -195,6 +195,18 @@
+	bic	r0, r0, #SCR_FW_BIT
+	stcopr	r0, SCR
+#endif
+
```
The `el3_entrypoint_common` macro is also used by BL1. This code would be better placed in `sp_min_entrypoint`.

etienne-lms replied:
right, my fault, sorry. I see 2 ways of implementing this, which do you prefer ?
- either add the sequence at 2 places in bl32/sp_min/aarch32/entrypoint.S,
in `sp_min_entrypoint` and `sp_min_warm_entrypoint`.
- or condition this to `#if defined(IMAGE_BL32) && SP_MIN_WITH_SECURE_FIQ`

danh-arm replied:
I think the latter is fine. It's better to avoid duplicated code.

etienne-lms commented on bl32/sp_min/aarch32/entrypoint.S:
```diff
@@ -166,6 +166,44 @@ func handle_smc
+	smcc_save_gp_mode_regs
+
+	/*
+	 * ARMv7 architectures need to clear the exclusive access when
```
ok.

etienne-lms commented:
Rebased, comments addressed.
Updated the doc change:
```patch
+   Platforms that enable FIQ handling in SP_MIN shall implement
+   the api `sp_min_plat_fiq_handler()`.
```

soby-mathew commented on include/common/aarch32/el3_common_macros.S:
```diff
@@ -195,6 +195,18 @@
 	cps	#MODE32_mon
 	isb
 
+#if defined(IMAGE_BL32) && SP_MIN_WITH_SECURE_FIQ
```
nit: Would prefer to avoid SP_MIN build flags in el3_common macros. Perhaps an additional argument to this macro (_route_fiq?) and SP_MIN entry point can encode this argument based on SP_MIN_WITH_SECURE_FIQ flag ?

Please dont do the rework yet. Let me confirm that the suggestion is alright.

etienne-lms replied:
ok. i see your idea. 
If you feel this implementation better suites the atf generic way, I am fine.

A drawback a see is to clearly define what such `_route_fiq` argument would mean on AArch64 version of the macro `el3_entrypoint_common`. And digressions on fiq/irq/where-are-the-secure-interrupts.

What about a new directive `BL32_ROUTE_SECURE_FIQ=undefined|0|1` that would be set from `SP_MIN_WITH_SECURE_FIQ` in the **make_helpers/default.mk**. **el3_common_macros.S** would rely on `BL32_ROUTE_SECURE_FIQ` not `SP_MIN_WITH_SECURE_FIQ`.

etienne-lms replied:
Following discussion from https://github.com/ARM-software/arm-trusted-firmware/pull/977#pullrequestreview-46814003, i will move back this SCR init so that it remains in the sp_min in **bl32/sp_min/aarch32/entrypoint.S**.

etienne-lms commented:
Thanks for the feedback. Yes, i will squash and update the P-R.

soby-mathew commented on Makefile:
```diff
@@ -455,6 +455,7 @@ $(eval $(call assert_boolean,USE_COHERENT_MEM))
 $(eval $(call assert_boolean,USE_TBBR_DEFS))
 $(eval $(call assert_boolean,WARMBOOT_ENABLE_DCACHE_EARLY))
 $(eval $(call assert_boolean,ENABLE_SPE_FOR_LOWER_ELS))
+$(eval $(call assert_boolean,SP_MIN_WITH_SECURE_FIQ))
```
Sorry, I didn't notice this before. This check has to go into sp_min.mk.

soby-mathew commented on Makefile:
```diff
@@ -495,6 +496,7 @@ $(eval $(call add_define,USE_COHERENT_MEM))
 $(eval $(call add_define,USE_TBBR_DEFS))
 $(eval $(call add_define,WARMBOOT_ENABLE_DCACHE_EARLY))
 $(eval $(call add_define,ENABLE_SPE_FOR_LOWER_ELS))
+$(eval $(call add_define,SP_MIN_WITH_SECURE_FIQ))
```
Same comment as above.

soby-mathew commented on make_helpers/defaults.mk:
```diff
@@ -119,6 +119,10 @@ SPD				:= none
+# Flag to allow SP_MIN to handle FIQ interrupts in monitor mode. The platform
+# port is free to override this value. It is default disabled.
+SP_MIN_WITH_SECURE_FIQ 		:= 0
+
```
This has to go to `sp_min.mk`.

davidcunado-arm commented:
@etienne-lms 
Need to re-base following doc migration to .rst format. Also, I think there are some comments from Soby that still need addressing.
Thanks!

danh-arm commented:
jenkins: test this please
